sdks/java/harness/src/main/java/org/apache/beam/fn/harness/FnApiDoFnRunner.java

```diff
@@ -941,20 +1047,59 @@ private Progress getProgress() {
           convertSplitResultToWindowedSplitResult.apply(result, watermarkAndState.getValue());
     }
 
+    List<BundleApplication> primaryRoots = new ArrayList<>();
```

Comment:
This also takes care of self checkpoint, right?

Reply:
Yes. As shown in the test the self checkpoint will checkpoint for the "remaining" windows as well.

I would like to leave it as is until we can build consensus around what we should do in these use cases.